doc/format-run.md

```diff
@@ -0,0 +1,216 @@
+2^32 - 2^16 - 1 = 4294901759, though this limit could also be rounded down in
+a public API.
+
+## Blobs file
+
+Each key/operation in the key/operations file optionally contains a blob
+references. A blob reference is of course a reference to a blob value, which
+is simply a sequence of bytes. These blob values are stored in the blobs file.
+
+The representation of a blob reference is a 64bit byte file offset and a 32bit
+byte length. This identifies a span of bytes within this blob file.
+
+The file format is simple: the concatenation of all the blob values, in the
+same order as the key/operations that refer to the blobs (which is sorted by
+key). There is no additional padding or alignment.
+
+This simple format supports writing out the LSM run incrementally, as entries
+are added sequentially in key order. Blobs are retrieved simply by reading the
+blob span into memory.
+
+To keep things simple and uniform, there is no special case for the empty
+sequence of blobs: it is represented by an empty blobs file.
+
+## Bloom filter file
+
+The format of this file consists of a header, followed by the bloom filter bit
+vector.
+
+The first 4 bytes (32bit) are a format identifier / format version. This
+determines the format of the rest of the header and file.
+
+The remainder of the header for format 0 consists of:
+ 1. The hash function count (32bit)
+ 2. The bit size of the filter (32bit)
+ 3. A boolean to indicate endianness, 0 little, 1 big (32bit)
+
+The family of hash functions to use is implied by the format version.
+
+The filter bit vector itself is organised as a whole number of 32bit words.
+It follows immediately after the header. The endianness from the header
+indicates the endianness of the 32bit words.
+
+## Index file
+
+The index maps keys to the number of the disk page(s) in the key/operation file
+that _could_ contain this key.
+
+That is, the index says where the key could possiblly be, and thus what pages
+to load from disk to search for the key. The index does not say whether a key
+exists or not (the bloom filter does this, albeit probabilistically, with false
+positives).
+
+For the case of large values that require multiple pages, the index must report
+the span of disk pages that the key/operation may be on. This allows the whole
+range of pages to be read from disk in one go. (An alternative design would be
+for the first page to indicate that it is a multi-page value, and then require
+a second round of disk I/O to read in the remaining pages.)
+
+The design is intended to be somewhat extensible by allowing for different
+choices of index representation, to suite different kinds of key and key
+properties. In particular initially we will support a compact index type.
+For more general purpose applications one can forsee the need for a more
+ordinary index type.
+
+The type of index in use is metadata that must be known for the whole LSM tree.
+It does not vary per run.
```

Comment:
Maybe we need to be precise about our terminology: do we mean the same thing by "LSM tree" and "table handle"?

Reply:
changed to table handle